### ISSUE TITLE
auto_ptr is deprecated

### ISSUE THREAD

**Issue by rcseacord** [zcash/zcash#1234]
C++11 has deprecated std::auto_ptr in favor of std::unique_ptr. The choice of smart pointer will depend on your use case and your requirements, with std::unique_ptr with move semantics for single ownership that can be used inside containers (using move semantics) and std::shared_ptr when ownership is shared.

You should try to use the smart pointer that best fits the situation, choosing the correct pointer type provides other programmers with insight into your design.

For example:

In file included from ./boost/smart_ptr/shared_ptr.hpp:28:0,
                 from ./boost/shared_ptr.hpp:17,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/detail/shared_count.hpp:395:33: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     explicit shared_count( std::auto_ptr<Y> & r ): pi_( new sp_counted_impl_p<Y>( r.get() ) )
                                 ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp:242:65: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
 template< class T, class R > struct sp_enable_if_auto_ptr< std::auto_ptr< T >, R >
                                                                 ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp:441:31: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     explicit shared_ptr( std::auto_ptr<Y> & r ): px(r.get()), pn()
                               ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp:454:22: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     shared_ptr( std::auto_ptr<Y> && r ): px(r.get()), pn()
                      ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp:520:34: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     shared_ptr & operator=( std::auto_ptr<Y> & r )
                                  ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp:529:34: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
     shared_ptr & operator=( std::auto_ptr<Y> && r )
                                  ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~
In file included from ./boost/shared_ptr.hpp:17:0,
                 from ./boost/filesystem/path.hpp:29,
                 from libs/filesystem/src/path.cpp:26:
./boost/smart_ptr/shared_ptr.hpp: In member function ‘boost::shared_ptr<T>& boost::shared_ptr<T>::operator=(std::auto_ptr<_Up>&&)’:
./boost/smart_ptr/shared_ptr.hpp:531:38: warning: ‘template<class> class std::auto_ptr’ is deprecated [-Wdeprecated-declarations]
         this_type( static_cast< std::auto_ptr<Y> && >( r ) ).swap( *this );
                                      ^~~~~~~~
In file included from /usr/include/c++/6/bits/locale_conv.h:41:0,
                 from /usr/include/c++/6/locale:43,
                 from ./boost/filesystem/path_traits.hpp:29,
                 from ./boost/filesystem/path.hpp:25,
                 from libs/filesystem/src/path.cpp:26:
/usr/include/c++/6/bits/unique_ptr.h:49:28: note: declared here
   template<typename> class auto_ptr;
                            ^~~~~~~~

**Comment by daira** [comment]
The uses of ``auto_ptr`` are within boost. Here's the upstream bug: https://svn.boost.org/trac/boost/ticket/11622

A comment on that ticket says that it "should be fixed"; the linked commit uses the approach to selectively disable the warnings that I suggested [here](https://github.com/zcash/zcash/issues/911#issuecomment-217049144)
(but in the boost headers themselves as @defuse suggested a little later on that ticket). So it seems that the boost devs have no immediate interest in switching away from ``auto_ptr``. Note that we disabled deprecation warnings ourselves globally, in https://github.com/zcash/zcash/pull/939/commits/ec3bd70b6d6ef8be901a5523a401e333c52e7bbe but see also #944.
